New article: your tech stack matters (#350)
subvisual/subvisual.com: src/posts/2020-06-20-tech-stack-matters.md

title: "Your tech stack matters"
author: luis-zamith
date: 2020-06-20
tags:
  - development
intro: Your about to start building a digital product? Here's some consideration to consider before defining your tech stack.
---

There's a common idea that when building a digital product, particularly online,
all tech stacks are created equally and regardless of what you choose you'll
live or die on something else. I believe this to be blatantly untrue, but maybe
not for the reasons you're thinking.

Here's the gist of it:

1. Some tech stacks add boilerplate, impeding rapid learning
2. More moving pieces equals more headaches
3. What your team is excited about matters
4. There are no silver bullets

This should be enough to give an idea of what I'm talking about, but let's dig a
bit deeper into each of these points. A disclaimer before we get going though,
this is my personal point of view and it sure has a lot of controversial ideas,
however after many years in the business of creating products I belive them to
be true. Proceed at your own peril.

## Some tech stacks add boilerplate, impeding rapid learning

Let's start looking at this from the programming language angle. You can build

> anything we pretty much any programming language, as long as it is Turing

Reviewer: ```suggestion
anything with pretty much any programming language, as long as it is Turing
```

complete, but there are languages that are a better fit for different types of
jobs. I personally don't think that the language itself is going to be the
biggest pain, but you get what you paid for. I wouldn't recommend that you start
your business using COBOL, as I wouldn't recommend that you start it in Java
either, but it does boil down to personal preference.

Where I think it starts to really matter is when it comes to productivity and

> even more important speed of change. Full disclosure I'm a big fan of both Ruby

Reviewer: ```suggestion
even more importantly, speed of change. Full disclosure, I'm a big fan of both Ruby
```

and Elixir, and by extension have used Rails and Phoenix a lot. However, even

> with all it's know issues, Rails is still a huge productivity boost that I would

Reviewer: ```suggestion
with all its know issues, Rails is still a huge productivity boost that I would
```

be remiss not to recommend. As Ryan Singer (Head of Strategy at Basecamp) puts
it "it's a 10x multipler for them".

Phoenix is great in a lot of ways, but not as malleable as Rails, which in all
honesty is all you should care about when you're still trying to figure out what
your product should do. That's the main reason I am not a fan of "batteries not
included" type of frameworks like Sinatra in Ruby or Express in JS, which trick
you into thinking you're better off starting small with something that is not
"bloated", but ultimately make you reimplement pretty much everything (or import
it via libraries many times not controlled by the core team).

> You don't have to choose Rails, there are technical downsides too, like memory

Reviewer: ```suggestion
You don't have to choose Rails. There are technical downsides too, like memory
```

usage or being late to the interactive apps game (mitigated by webpacker,

> actioncable and stymulus), but when starting an online product prioritise for

Reviewer: ```suggestion
ActionCable and Stymulus), but when starting an online product prioritise for
```

> batteries included and easy to change, because trust me you will change, a lot.

Reviewer: ```suggestion
batteries included and easy to change, because trust me, you will change, a lot.
```

## More moving pieces equals more headaches

> Having way more parts to a tech stack that what you really need to validate your

Reviewer: ```suggestion
Having way more parts to a tech stack than what you really need to validate your
```

there's no good reason for the vast majority of new products to do it, yet it is
now extremely prevalent. Think of Conway's Law:

>  Any organization that designs a system (defined broadly) will produce a
>  design whose structure is a copy of the organization's communication
>  structure.

For a large company, with many different teams and departments, communication
across them is hard, therefore they design a system that makes that
communication hard as well. In this case, you have a Front End application that
only talks to the Back End via a well defined API (contract) and changes to that
contracts must be agreed upon by both parties. Moreover, both parties will
implements their own state management, routing, translations and even data
validations. Obviously there are ways to make these things less obvious and
(debatably) less painful, but as a small company, why go there in the first
place?

Again, I advocate for easy to change here, because remember that's what we are
prioritising. If you really need the interactivity in the client, which is a
resonable argument to make, use something like PJAX, Stymulus or LiveView (there

> are likely others), who allow you to handle mostly everything from your server

Reviewer: ```suggestion
are likely others), which allows you to handle mostly everything from your server
```

and remove the duplication of routing, state management, etc. You do lose the
ability to have the app work offline, which, let's face it, is a fringe scenario
anyway and in most (if not all) cases not relevant at this stage.

Every line of code is a potential bug, the ideal system does all you want with
zero lines of code, therefore you want to minimize the amount of code needed to
achieve your goals. Not using a Front End framework is a good start, but it can

> go to infrastructure to, why not get rid of Docker, Kubernetes, Kafka,

Reviewer: ```suggestion
go to infrastructure too. Why not get rid of Docker, Kubernetes, Kafka,
```

> Tenserflow and so on, and so on. All of these tools have a time and a place to

Reviewer: ```suggestion
Tensorflow,, and so on? All of these tools have a time and a place to
```

> be used, don't get me wrong, I just think it should be top of mind for the great

Reviewer: ```suggestion
be used, don't get me wrong. I just think it should be top of mind for the great
```

Reviewer: I don't undertand what "should be the top of mind" is suposed to mean

Author: It should be "shouldn't"

Reviewer: so, "shouldnt' be *on the of anyone's find*" or something like that? still feels incomplete as it is

majority of products just getting started.

Ah, and for those that took my ideal system to be a serverless one, that's
simply moving lines of code from a place under your control to a place you don't
control. That's not ideal at all, especially if your core business is that

> system. Same with no code platforms, for that matter. I'm sure you can build

Reviewer: ```suggestion
system. Same with no-code platforms, for that matter. I'm sure you can build
```
